README.md
# Ruiby

[![Build Status](https://travis-ci.org/glurp/dsl-gtk.svg?branch=master)](https://travis-ci.org/glurp/dsl-gtk)
[![Gem Version](https://badge.fury.io/rb/Ruiby.png)](http://badge.fury.io/rb/Ruiby)
[![AppVeyor](https://ci.appveyor.com/api/projects/status/y6pqyq79ybcmj9ye?svg=true)](https://ci.appveyor.com/project/Glurp/dsl-gtk/branch/master)

A DSL for building simple GUI ruby application.
Based on gtk.


Resources
==========


Code: http://github.com/glurp/Ruiby

Doc: [Reference+Exemples.](https://rawgithub.com/glurp/Ruiby/master/doc.html)

Gem : https://rubygems.org/gems/Ruiby

Based on Ruby-Gmome2 :
* [Sources, Issues](https://github.com/ruby-gnome2/ruby-gnome2)
* [API](http://ruby-gnome2.osdn.jp/hiki.cgi?Ruby%2FGTK)


Status
======

NEW : 3.24.0 !!   28-11-2017 : Linear Gradient on canvas draw surfaces primitives

Now, plotters and dashboard can have beautiful background :)



TO DO  :

* [x] improve graphics quality on canvas : linear gradient
* [x] improve graphics quality on canvas : radial  gradient
* [x] improve graphics quality on canvas : transparency
* [ ] refactoring samples demos with last improve: dynvar/autoslot...
* [ ] resolve 100% gtk3 deprecated warning
* [ ] complete rspec => 99% coverage ?

Abandoned :

* [x] gadget API





Installation
============
1) system

Install Ruby 2.x  #  x>1


2) install Ruiby
(```gem install Ruiby``` will install ruby-gtk3 which install gtk3 libs)
```
> gem install Ruiby
```

Test it:

```
> ruiby_demo             # check good installation with gtk3 (default)
> ruiby_sketchi          # write and test ruiby code
```


Here a working gem config on windows (25-Sept-2016, Ruby  2.3.3p222) :
```
  did_you_mean-1.0.0
  pkg-config-1.2.3
  native-package-installer-1.0.4
  cairo-1.15.9
  glib2-3.1.8
  gobject-introspection-3.1.8
  gio2-3.1.8
  atk-3.1.8
  cairo-gobject-3.1.8
  pango-3.1.8
  gdk_pixbuf2-3.1.8
  gdk3-3.1.8
  rsvg2-3.1.8
  gtk3-3.1.8
  Ruiby-3.23.0
  gtksourceview3-3.1.8
```


Usage
======
DSL is usable via inherit, include, Ruiby.app bloc, or one-liner command.

By inherit:

```ruby
class Application < Ruiby_gtk
    def initialize(t,w,h)
        super(t,w,h)
    end
	def component()
	  stack do
		...
	  end
	end
	.....your code....
end
Ruiby.start { Win.new("application title",350,10) }

```

By include, calling ruiby-component() :

```ruby
class Win < Gtk::Window
	include Ruiby
    def initialize(t,w,h)
        super()
		add(@vb=VBox.new(false, 2))
		....
    end
	def add_a_ruiby_button()
		ruiby_component do
			append_to(@vb) do
				button("Hello Word #{@vb.children.size}") {
					add_a_ruiby_button()
				}
			end
		end
	end
end
Ruiby.start { Win.new("application title",350,10) }
```

Autonomous DSL, for  little application (most of demo in samples/ are
done with this pattern) :

```ruby
require  'Ruiby'
Ruiby.app do
	stack do
		. . .
	end
end
```
And, for very little application ('~' are replaced by guillemet):

```ruby

> ruiby   button(~Continue ? ~) "{  exit!(0) }"
> ruiby   fields([%w{a b},%w{b c},%w{c d}]) { "|a,b,c|" p [a,b,c] if a; exit!(a ?0:1) }
> ruiby -width 100  -height 300 -title "Please, select a file" \
             l=list(~Files :~);l.set_data Dir.glob(~*~) ;  \
             buttoni(~Selected~) { puts l.selection ; exit!(0) } ;\
			 buttoni(~Annul~) { exit!(1) }

```

Require
=======
Simple usage with gtk3 :

```ruby
require 'Ruiby'
```


Usage with Event Machine: load event-machine before Ruiby :

```ruby
require 'em-proxy'
require 'Ruiby'
```

Warning : EM.run is done when starting mainloop, after creation of window(s).
So, if you need initialization of event-machine callback, do it in component(), in a after(0):

```ruby
Ruiby.app do
  ....
  after(0) { EventMachine::start_server().. { ... } }
end
```

See samples/spygui.rb, for example of GUI with EM.


Threading
=========
Ruiby does not have confidence in gtk multi threading, so all Ruiby commands must be done in
main thread context. A Ruiby delegate is provided in Kernel module for support multi-threading

A Queue is polled by main-window thread :
* main window poll Queue , messages are proc to be instance_eval() in the main window context
* everywhere, a thread can invoke ```invoke_gui {ruiby code}```. this send to the main queue the proc,
   which will be evaluated asynchronously

instance_eval is avoided in ruiby. He is used only for thread invoker : gui_invoke().

```ruby
require_relative '../lib/Ruiby'
class App < Ruiby_gtk
    def initialize
        super("Testing Ruiby for Threading",150,0)
		threader(10)
		Thread.new { A.new.run }
    end
	def component()
	  stack do
		sloti(label("Hello, this is Thread test !"))
		stack { @lab=stacki { } }
	  end
	end # endcomponent

end
class A
	def run
 		loop do
		 	sleep(1) # thread...
			there=self
			gui_invoke { append_to(@lab) { sloti(
					label( there.aaa )  # ! instance_eval on main window
			)  } }
		end
	end
	def aaa() Time.now.to_s  end
end

Ruiby.start { App.new }

```


Observed Object/Variable
========================

Dynamic variable
----------------
Often, a widget (an entry, a label, a slider...) show the value of a ruby variable.
each time a code modify this variable, it must modify the widget, and vice-versa...
This is very tiring :)

With data binding, this notifications are done by the framework

So ```DynVar``` can be  used for representing a value variable which is dynamics, ie.
which must notify widgets which show the variable state.

So we can do :
```ruby
  foo=DynVar.new(0)
  entry(foo)
  islider(foo)
  ....
  foo.value=43  
  ....
```

That works ! the entry and the slider will be updated.

A move on slider will update foo.value and the entry.
Idem for a key in the entry : slider and foo.value will be updated.

if you want to be notified for your own treatment, you can observe a DynVar :

```ruby
  foo.observ { |v| @socket.puts(v.to_s) rescue nil }
```

Here, a modification of foo variable will be send on the network...

Warning !! the block will always be executed in the main thread context (mainloop gtk context).
So DynVar is a resource internal to Ruiby framework.

Widget which accept DynVar are : entry, ientry, islider, label, check_button,

```
must be extend to button, togglebutton, combo, radio_button ... list, grid,...
```


Dynamic Object
--------------

Often, this kind of Dyn variables are members of a 'record', which should be organized by an
Ruby Object (a Struct...)

So ```DynObject``` create a class, which is organized by a hash  :
* packet of variable name
* put initial value for each
* each variable will be a DynVar

```ruby
  FooClass=make_DynClass("v1" => 1 , "v2" => 2, "s1" => 'Hello...')
  foo=FooClass.new( "s1" => Time.now.to_s ) # default value of s1 variable is replaced
  ...
  label(" foo: ") ; entry(foo.s1)
  islider(foo.v1)
  islider(foo.v2)
  ....
  button("4x33") { Thread.new { foo.s1.value="s4e33" ; foo.v2.value=33 ; foo.v1.value=4} }
  ....
```

Dynamic Stock Object
--------------------
DynObject can be persisted to file system : use ```make_StockDynObject```, and
instantiate with an object persistent ID

```ruby
  FooClass=make_StockDynClass("v1"=> 1 , "v2" => 2, "s1" => 'Hello...')
  foo1=FooClass.new( "foo1" , "s1" => Time.now.to_s )
  foo2=FooClass.new( "foo2" , "s1" => (Time.now+10).to_s )
  ....
  button("Exit") { ruiby_exit} # on exit, foo1 and foo2 will been saved to {tmpdir}/<$0>.storage  
  ....
```
`make_StockDynObject` do both : Class creation **and** class instantiation.

```ruby
  foo=make_StockDynObject("v1"=> 1 , "v2" => 2, "s1" => 'Hello...')
  ....
  button(foo.s1) { foo.s1.value= prompt("new S1 value ?")}
  button("Exit") { ruiby_exit} # on exit, foo1 and foo2 will been saved to {tmpdir}/<$0>.storage  
  ....
```

Component
=========
Ruiby is not really object-orented : most of DSL words are simple method in Ruby_dsl module.

Sometime, this is not good enough :
* when a component must have many specific methods
* when component have (model) state : variable member must be used

So Component concept has been added (Fev 2016).It authorize to define a
class, child of AbstractComponent, which can be used by a DSL Word.

Components code seem very close to a Ruiby window : free constructor,
define ```component()``` method for draw the widgets

Create a component:
```ruby
class AAA < AbstractComposant
   def initialize(name)
      @name= name
      @state=1
   end
   def component()
    framei("Component Comp:#{@name}") do
      label_clickable("B#{@name}...") { @state=2 }
      entry(@name,4)
    end
   end
   def get_state() @state end
end
```

Define a word which instantiate a component of class AAA:
```ruby
module Ruiby_dsl
  def aaa(*args)
    c=install_composant(self,AAA.new(*args))
  end
end
```

Use the component:
```ruby
        c=nil
        stack {
           c=aaa "foo"
           flowi { aaa 1; aaa 2 }
        }
        button("?") { alert( c.get_state() ) }
```

A demo is at ```samples/composant.rb```.

TO-DO:
* Canvas and Plot must be converted to Component, soon :)
* Define ```destroy()```
* Hook for auto-generate DSL word
* Test Stock, Dynvar, threading,
* Tests, tests, test...


License
=======
Ruiby                   : LGPL, CC BY-SA

fafamfam rasters images : CC Attribution 4.0 http://www.famfamfam.com/

Crystal Clear icon set  : LGPL

Farm Fresh icon set     :  CC Attribution 3.0 License http://www.fatcow.com/free-icons

Exemples
========
See samples in "./samples" directory (run all.rb)
See at end of Doc reference : [Ex.](https://rawgithub.com/glurp/Ruiby/master/doc.html#code)


CHANGELOG.txt
3.24.6
    video.rb : debug
    pipeplot.rb : update
    pingpongplot.rb : comment
    ex.rb : concole correction
    animtext.rb : title
    ruiby_dsl3.rb : validation
    Ruiby.gemspec : add pango and cairo as dependency
    canvas.rb : text bg position and special traitment for exception in drawing
    README.md : gradient
    ruiby_dsl3.rb : color parse take care of Array for color name
    plot.rb : no bordures for plot zone
    README.md : gradiant
    test.rb : gradiant on canvas demo
    README.md : gradient
    README.md : orthographe
3.24.0 : 2017-11-28 19:12:41 +0100
    quadvideo.rb : q
    circle.rb : bg with gradient
    ruiby_dsl3.rb : error() with exception for video()
    plot.rb : correction color
    list_grid.rb : no unselect on set_data
    canvas.rb : gradiant for draw_rectangle, draw_rounded_rectangle, drw_polygon
    README.md : gradient evolution, ruby-gnome2 references
    README.md : updates
    README.md : update dependencies
3.23.0 : 2017-09-25 17:43:54 +0200
    README.md : 3.23
    ruiby_default_dialog3.rb : dialog synced
    plot.rb : draw polygon 
    commands.rb : define trace()
    canvas.rb : declare font and size for drawing text
    .travis.yml : trusty version for ubuntu execution
    pixmap.rb : gtk3
    form_fields.rb : trxt_area_dyn correction
    commands.rb : create_log_widget : log applicationato can be embedded
    getdownstat.rb : echo
3.22.0 : 2017-06-12 16:11:43 +0200
    doc.html : update
    appveyor.yml : ruby 2.3
    README.md : update
3.21.0 : 2017-06-02 13:11:54 +0200
3.20.0 : 2017-06-02 13:07:36 +0200
    README.md : version 3.20
    dyn.rb : nl
    layouts.rb : add context to dialog container
    README.md : to gnome-gtk 3.1.3
3.19.1
    Ruiby.rb : gtk3 only
3.19.0 : 2016-08-26 23:56:10 +0200
3.18.0 : 2016-08-26 23:55:33 +0200
    Ruiby.gemspec : gtk3 dependecie
3.17.3
    ruiby_dsl3.rb : add reqire plot.rb
    canvas.rb : correction doc on cv.draw*
    Ruiby.rb : adde require default_dialog
    canvas.rb : fill drawing text tools
    Ruiby.rb : supression gtk2
    Ruiby.gemspec : updates
    Ruiby.rb : gtk_version with Gtk::Version.or_later?
3.17.0 : 2016-08-18 13:17:08 +0200
    README.md : update
3.16.0 : 2016-06-13 13:26:21 +0200
3.15.0 : 2016-06-10 13:33:20 +0200
3.14.0 : 2016-05-24 16:21:28 +0200
    icones.rb : window size
    windows.rb : stop event Machine at window destroy
    README.md : update
3.13.0 : 2016-05-16 13:37:43 +0200
    test_layout.rb : left/right test
    test.rb : refactoring notepad
    images.rb : coorected
    ruiby_dsl3.rb : ToolButton gtk3
    ruiby_default_dialog3.rb : propteSync() creation
    pixmap.rb : get_pixmap corrected for Label/Button/Image
    label_button_image.rb : image on Button/Image corrected, with tooltips
    commands.rb : log window with title
    Ruiby.rb : ruiby_require corrected
    Ruiby.gemspec : force 3.0.7 gtk version
3.12.0 : 2016-04-21 15:16:02 +0200
    test.rb : edit all sources
    Ruiby.gemspec : ruby version >= 2.1
    README.md : update
3.12.0 : 2016-04-21 15:12:41 +0200
3.11.0 : 2016-04-21 14:55:22 +0200
    cal.rb : require Ruiby
    ruiby_threader.rb : get pending event queue size
    layouts.rb : fremei does stacki
    canvas.rb : no traces
    Ruiby.rb : test ruby version
3.10.0 : 2016-02-17 22:38:03 +0100
    pipeplot.rb : wait option
    canvas.rb : plot tracking draw
    README.md : update
    README.md : update
3.9.0 : 2016-02-04 23:24:23 +0100
    layouts.rb : dialog validation
    label_button_image.rb : show_source creation
    form_fields.rb : combo corrections
    README.md : update
    README.md : update
    ruiby_default_dialog.rb : orthographe
    layouts.rb : dialog closure with dialog widget as agument
3.8.0 : 2015-12-24 16:29:15 +0100
    sketchi.rb : list of icone 
    ruiby_default_dialog3.rb : edit xith callback
    ruiby_default_dialog.rb : edit with callback
    editor.rb : edit with callback
    pixmap.rb : traces
    layouts.rb : panel_async creation
    form_fields.rb : toggle button and dyn value
    canvas.rb : scale
    dyn_var.rb : destroy creation
    layouts.rb : panel_async for dialog without buttons
    form_fields.rb : progress bar creation
    canvas.rb : comments
    Rakefile.rb : project name diff of directory name
    README.md : updated
3.7.0 : 2015-10-16 23:43:45 +0200
3.6.0 : 2015-10-16 23:41:42 +0200
3.5.0 : 2015-10-16 23:38:18 +0200
3.4.0 : 2015-10-16 23:33:55 +0200
    make_doc.rb : plot_curve as example
    gadget_trading.rb : archives
    README.md : october news
    Rakefile.rb : project name
    README.md : update
3.2.7
    doc.html : update
    canvas.rb : bar in plot
    gadget_trading.rb : recreation
    canvas.rb : plot multicurve
    Gemfile : gtksourceview3 referenced
    label_button_image.rb : slider with dyn variable,, decimal option
    sketchi.rb : list stock update for gtk3
    ruiby_dsl3.rb : color convertions for gtk3
    canvas.rb : draw_* complements, debugs
    test.rb : drawing canvas demo
    README.md : update gtk3.0.3
3.2.0 : 2015-09-16 21:32:15 +0200
3.1.0 : 2015-09-16 21:29:55 +0200
    README.md : updae for 3.0.3
    test.rb : icon names have cange for gtk3
    fractal.rb : color naming for gtk3
    ruiby_dsl3.rb : rgba() correction...nott sure, for gtk3
    ruiby_default_dialog3.rb : ColorChooserDialog in gtk3
    ruiby_default_dialog.rb : qcolorChooserDialog have changeing name in gtk3
    pixmap.rb : ivon names changes in gtk3
    layouts.rb : icons for notebook for gtk3
    canvas.rb : :q
    Ruiby.gemspec : dependancy gtksousrceview3 on 2.2.5
    test.rb :  update 3.0.0
    paned.rb :  update 3.0.0
    hello.rb :  update 3.0.0
    cal.rb :  update 3.0.0
    windows.rb :  update 3.0.0
    ruiby_dsl3.rb :  update 3.0.0
    pixmap.rb :  update 3.0.0
    menus_popup.rb :  update 3.0.0
    list_grid.rb :  update 3.0.0
    layouts.rb :  update 3.0.0
    form_fields.rb :  update 3.0.0
    Ruiby.rb :  update 3.0.0
    README.md : update 3.0.0
1.33.0 : 2015-05-26 19:28:12 +0200

lib/ruiby.rb
###############################################################################################
#            ruiby.rb :   A DSL for Ruby/Gui
#                         Gtk based. Should will support SWT , Qt, WinForms ?...
###############################################################################################
# Creative Commons BY-SA :  Regis d'Aubarede <regis.aubarede@gmail.com>
# LGPL
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
# 
require 'tmpdir'
require 'thread'
require 'pathname'
require 'gtk3' if ! defined?(Gtk) # caller can preload  gtk3/gtk2, at his own risk...

#require 'gtksourceviewX' # done by source_editor() tag, so only if source edit is needed

module Ruiby
  DIR = Pathname.new(__FILE__).realpath.dirname.to_s
  VERSION = IO.read(File.join(DIR, '../VERSION')).chomp
  GUI='gtk'
  
  # Update gui while event is pending
  # to be use if current algorithme is long and user want see gui update
  # log long list of message to screen...
  #
  def self.update() 
    Gtk.main_iteration while Gtk.events_pending?  
  end
  def self.gtk_version(major)
    Gtk.check_version(major, 0, 0)==nil
  end
  ###########################################################
  #                S t o r a g e 
  ###########################################################
  
  # persistent stock a value associated to a name
  # dictionary is serialised (Marshalling) to a default file
  # file is /tmp/$0.storage
  def self.stock_put(name,value)
    db="#{Dir.tmpdir}/#{File.basename($0)}.storage"
    data={}
    (File.open(db,"r") { |f| data=Marshal.load(f) } if File.exists?(db)) rescue nil
    data[name]=value
      File.open(db,"w") { |f| Marshal.dump(data,f) }
  end
  # read a value associated to a name from persistant storage
  def self.stock_get(name,default="")
    db="#{Dir.tmpdir}/#{File.basename($0)}.storage"
    data={}
    (File.open(db,"r") { |f| data=Marshal.load(f) } if File.exists?(db) )rescue nil
    data[name] || default
  end
  # clear persistant strorage
  def self.stock_reset()
    db="#{Dir.tmpdir}/#{File.basename($0)}.storage"
    File.delete(db) if File.exists?(db)
  end
  ###########################################################
  #                start Ruiby application
  ###########################################################
  
  # start ruiby.
  # Usage: Ruiby.start { Win.new() }
  # One shot :reloading of source can be done, block wjile not be evaluated
  #
  # Thread.abort_on_exception, BasicSocket.do_not_reverse_lookup, 
  # trap('INT') are settings
  def self.start(&bloc)
    return if defined?($__MARKER_IS_RUIBY_INITIALIZED)
    $__MARKER_IS_RUIBY_INITIALIZED = true
    $stdout.sync=true 
    $stderr.sync=true 
    Thread.abort_on_exception = true  
    BasicSocket.do_not_reverse_lookup = true if defined?(BasicSocket)
    trap("INT") { exit!(0) }
    Gtk.init
    yield
    secure_main()	
  end
  
  # Start ruiby with a main loop which trap all error :
  # A Exception in a callback will not kill the process.
  # Gtk Exception are trapped, so process should not exited by ruiby fault!
  def self.start_secure(&bloc)
    return if defined?($__MARKER_IS_RUIBY_INITIALIZED)
    $__MARKER_IS_RUIBY_INITIALIZED = true
    $stdout.sync=true 
    $stderr.sync=true 
    Thread.abort_on_exception = true  
    BasicSocket.do_not_reverse_lookup = true if defined?(BasicSocket)
    trap("INT") { exit!(0) }
    Gtk.init
    yield
    secure_main()	
  end
  
  # Direct acces to Ruiby DSL
  # config can contain :title, :width, :height
  # Warning ! bolc is invoked vy instance_eval on window
  # def action() puts "Hello..." end
  # Ruib.app {
  #    stack do button("test") { action() } end
  # }
  def self.app(config={},&blk)
    $blk=blk
    klass = Class.new Ruiby_gtk do
      def initialize(title,w,h)
        super
        threader(10)
      end
    end
    klass.send(:define_method,:component,&blk)
    klass.send(:chrome,config[:chrome]) if config.has_key?(:chrome)
    start_secure { 
      w=klass.new(config[:title] || "",config[:width] ||600,config[:height] ||600) 
        w.send(:chrome,config[:chrome]) if config[:chrome]
    }
  end
  def self.set_last_log_window(win)
    @last_log=win
  end
  def self.destroy_log()
    return unless @last_log  && ! @last_log.destroyed?
    @last_log.destroy() rescue nil
    @last_log=nil
  end
end


#require_relative 'ruiby_gtk/ruiby_default_dialog.rb'
require_relative(Ruiby.gtk_version(2) ? 'ruiby_gtk/ruiby_dsl.rb' : 'ruiby_gtk/ruiby_dsl3.rb')
require_relative 'ruiby_gtk/ruiby_threader.rb'
require_relative 'ruiby_gtk/windows.rb'
require_relative 'ruiby_gtk/editor.rb'
require_relative 'ruiby_gtk/systray.rb'

Dir.glob("#{Ruiby::DIR}/plugins/*.rb").each do |filename| 
  autoload(File.basename(filename).split(".")[0].capitalize.to_sym, filename) 
end

module Kernel
  # do a gem require, and if fail, try to load the gem from internet.
  # asking  permission is done for each gem. the output of 'gem install'
  # id show in ruiby log window
  def ruiby_require(*gems)
    w=Ruiby_dialog.new
    gems.flatten.each do|gem| 
      begin
        require gem
      rescue LoadError => e
        rep=w.ask("Loading #{gems.join(', ')}\n\n'#{gem}' package is missing. Can I load it from internet ?")
        exit! unless rep
        Ruiby.update
        require 'open3'
        w.log("gem install  #{gem} --no-ri --no-rdoc")
        Ruiby.update
        Open3.popen3("gem install  #{gem} --no-ri --no-rdoc") { |si,so,se| 
          q=Queue.new
          Thread.new { loop {q.push(so.gets) } rescue p $!; q.push(nil)}
          Thread.new { loop {q.push(se.gets) } rescue p $!; q.push(nil)}
          str=""
          while str
            timeout(1) { str=q.pop } rescue p $!
            (w.log(str);str="") if str && str.size>0
            w.log Time.now
            Ruiby.update						
          end
        }
        w.log "done!"
        Ruiby.update
        Gem.clear_paths() 
        require(gem) 
        w.log("loading '#{gem}' ok!")
        Ruiby.update
      end		
    end
    w.destroy()
    Ruiby.update
  end
  # run gtk mainloop with trapping gtk/callback error
  # used by sketchi.rb, not good
  # see Ruiby.secure_main { }
  # if EventMachine is present, do loop with wixed EM/main_iteration
  def secure_main()
    if defined?(EM)
        puts "Ruiby mainloop: EM is detected, mainloop while be mixed EM/Gtk"
        EM::run do
          give_tick = proc do
            begin
              Gtk::main_iteration
              EM.next_tick(give_tick); 
            rescue Exception => e
              exit!(0) if e.to_s=="exit"
              $__mainwindow__.error("Error GTK : "+e.to_s + " :\n     " +  e.backtrace[0..10].join("\n     "))
            end
          end
          give_tick.call
        end
    else
      begin 
        Gtk.main 
        exit!(0)
      rescue Exception => e
        if e.to_s=="exit"
          $__mainwindow__.error("Error, see STDERR in console...")
        else
          $__mainwindow__.error("Error GTK : "+e.to_s + " :\n     " +  e.backtrace[0..10].join("\n     "))
        end
      end while true
    end
  end
  
  #######################################
  # C=hash_to_class({a: 11, b: 22})
  # o=C.new ;
  # p o.keys ;p o.a ; p o.b        # => 11,22
  # o.a=1 ;o.b=2 ;p o.a ; p o.b    # => 1 ,2
  #
  # o=C.new  a: 10,b: 20
  # o=C.new ;
  # p o.keys ;p o.a ; p o.b        # => 10,20
  #
  # o=C.new  a: 10
  # o=C.new ;
  # p o.keys ;p o.a ; p o.b ;      # ==> 10,22
  #
  def hash_to_class(h)
    Class.new() do
      def initialize(x=nil)
        self.def_init()
        @values= @values.merge(x) if x
      end
      define_method(:def_init) { @values= h.dup } 
      define_method(:keys) {  h.keys }
      h.each do |key,value|
        define_method(key) { @values[key] }
        define_method("#{key}=") { |x| @values[key]=x }
      end
    end
  end  
end
